File: plugin/suite/roster/orchestration/src/roster_manifest.py
"""Load and validate a roster package manifest (PP-FR-2).

A roster package declares itself with a `roster.json` at its root. The selector
reads it to find the catalog, the routing configuration, the role definitions,
and the shared policy directory. **The Agentic SDLC kernel never reads this
file** — that is the whole reason it is a sibling of `provider.json` rather than
an extension of it (OD-5): `provider.json`'s key set is closed in two
independent implementations, one of them the kernel this work is constrained not
to touch.

Two decisions are load-bearing here and are cheap to reverse by accident, so
they are stated rather than left to the schema:

**Unknown `schema_version` is rejected, not ignored** (OD-11). That decision
declined a `platform_compatibility` window, accepting that a manifest written
against different selector semantics fails however its differences happen to
present. Rejecting an unrecognised version is the mitigation adopted with it,
and it is the only signal a mismatched manifest gets.

**Path containment is enforced here rather than borrowed.** The logic is ported
from the kernel's `provider_resource()` (`kernel/agentic_sdlc/__init__.py:159-169`)
rather than imported, because `test_kernel_boundary.py:76-95` forbids importing
kernel code from `roster/` and that guard must keep passing. It is deliberately
*not* `glob_containment.py`, which answers a different question — glob-language
subset containment for `exclude_paths` shadowing, not filesystem path escape.
Reaching for it here would produce a check that compiles, passes its own tests,
and contains nothing.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

MANIFEST_FILENAME = "roster.json"
SUPPORTED_SCHEMA_VERSIONS = (1,)

_REQUIRED_PATH_FIELDS = ("catalog", "routing", "role_root", "shared_policy_root")
_REQUIRED_FIELDS = ("schema_version", "id", "version", *_REQUIRED_PATH_FIELDS)


class RosterManifestError(ValueError):
    """A roster manifest is missing, malformed, or declares an unusable path."""


@dataclass(frozen=True)
class RosterManifest:
    """A validated manifest with every declared path already resolved."""

    root: Path
    manifest_path: Path
    id: str
    version: str
    catalog: Path
    routing: Path
    role_root: Path
    shared_policy_root: Path


def default_roster_root() -> Path:
    """This checkout's own roster, derived from the platform's location.

    Never routed through the `roster.root` setting: that setting's *default* is
    this value, so computing it from the setting would be circular.
    """
    return Path(__file__).resolve().parents[2]


def _resource(root: Path, value: Any, field: str, *, directory: bool) -> Path:
    """Resolve one declared path, rejecting anything that escapes `root`.

    `root` must already be `.resolve()`d by the caller — the containment check
    compares resolved paths, so an unresolved root both under-rejects (a symlink
    inside it looks like an escape) and reads confusingly.
    """
    if not isinstance(value, str) or not value:
        raise RosterManifestError(
            f"roster manifest field {field!r} must be a non-empty relative path"
        )
    candidate = (root / value).resolve()
    if not candidate.is_relative_to(root):
        # Also catches an absolute value: `Path("/a") / "/etc/passwd"` is
        # `Path("/etc/passwd")` under pathlib's join semantics, so the escape
        # surfaces here rather than at an is_absolute() guard. Tested
        # explicitly, because that is a quirk a later refactor could remove
        # without noticing it was load-bearing.
        raise RosterManifestError(
            f"roster manifest field {field!r} escapes its manifest directory: {value!r}"
        )
    if directory and not candidate.is_dir():
        raise RosterManifestError(
            f"roster manifest field {field!r} names a directory that does not exist: {value!r}"
        )
    if not directory and not candidate.is_file():
        raise RosterManifestError(
            f"roster manifest field {field!r} names a file that does not exist: {value!r}"
        )
    return candidate


def load_roster_manifest(root: Path) -> RosterManifest:
    """Read and validate `<root>/roster.json`.

    Fails closed and by name at every step (intent §7 C4): a missing manifest, a
    malformed one, an unknown schema version, a missing required field, and a
    path that escapes or does not exist each raise with the offending item
    named.
    """
    root = Path(root).expanduser().resolve()
    manifest_path = root / MANIFEST_FILENAME
    if not manifest_path.is_file():
        raise RosterManifestError(
            f"roster package at {root} is missing {MANIFEST_FILENAME}"
        )

    try:
        raw = json.loads(manifest_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as error:
        raise RosterManifestError(
            f"{manifest_path}: {MANIFEST_FILENAME} must contain JSON ({error})"
        ) from error
    if not isinstance(raw, dict):
        raise RosterManifestError(f"{manifest_path}: root must be a JSON object")

    missing = [field for field in _REQUIRED_FIELDS if field not in raw]
    if missing:
        raise RosterManifestError(
            f"{manifest_path}: missing required field(s): {', '.join(sorted(missing))}"
        )

    # OD-11's mitigation. Rejecting rather than ignoring is what converts the
    # most likely version mismatch from silent misbehaviour into an error
    # naming the manifest -- and with no compatibility window it is the only
    # such signal a roster author gets.
    version = raw["schema_version"]
    if version not in SUPPORTED_SCHEMA_VERSIONS:
        raise RosterManifestError(
            f"{manifest_path}: unsupported schema_version {version!r}; "
            f"this selector supports {', '.join(str(v) for v in SUPPORTED_SCHEMA_VERSIONS)}"
        )

    for field in ("id", "version"):
        if not isinstance(raw[field], str) or not raw[field]:
            raise RosterManifestError(
                f"{manifest_path}: field {field!r} must be a non-empty string"
            )

    return RosterManifest(
        root=root,
        manifest_path=manifest_path,
        id=raw["id"],
        version=raw["version"],
        catalog=_resource(root, raw["catalog"], "catalog", directory=False),
        routing=_resource(root, raw["routing"], "routing", directory=False),
        role_root=_resource(root, raw["role_root"], "role_root", directory=True),
        shared_policy_root=_resource(
            root, raw["shared_policy_root"], "shared_policy_root", directory=True
        ),
    )

File: internal/selector/rostermanifest.go
package selector

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// A roster package's manifest: how the platform learns a roster's layout
// instead of assuming one.
//
// `cadre select --roster <dir>` points the selector at a roster package. Which
// file in that package holds the catalog, and which holds the routing rules, is
// the package's business -- it declares them in roster.json. Resolving
// <dir>/catalog.yaml and <dir>/orchestration/routing.json directly works for
// exactly one roster: the one this repository ships, whose manifest happens to
// declare those paths.
//
// Ported from roster/orchestration/src/roster_manifest.py.

// RosterManifestFilename is the bootstrap name, and the one path that cannot
// itself be indirected: something has to know what the manifest is called in
// order to open it.
const RosterManifestFilename = "roster.json"

// SupportedRosterSchemaVersion is the only manifest schema this selector
// understands. A newer package is refused by version rather than being read
// with today's assumptions about what its fields mean.
const SupportedRosterSchemaVersion = 1

// RosterManifest is a validated manifest with every declared path resolved.
type RosterManifest struct {
	Root             string
	ManifestPath     string
	ID               string
	Version          string
	Catalog          string
	Routing          string
	RoleRoot         string
	SharedPolicyRoot string
}

// RosterManifestError is a manifest that is missing, malformed, or declares an
// unusable path.
type RosterManifestError struct{ message string }

func (e *RosterManifestError) Error() string { return e.message }

func rosterManifestErrorf(format string, args ...any) error {
	return &RosterManifestError{message: fmt.Sprintf(format, args...)}
}

// LoadRosterManifest reads and validates the manifest at root.
//
// A missing manifest is refused rather than defaulted. Falling back to this
// repository's own layout would mean a package that is not a roster package
// silently dispatches Cadre's roles -- the caller asked for one roster and got
// another, with a plan that looks entirely normal.
func LoadRosterManifest(root string) (*RosterManifest, error) {
	resolvedRoot, err := filepath.Abs(root)
	if err != nil {
		return nil, rosterManifestErrorf("roster package at %s cannot be resolved: %s", root, err)
	}
	if evaluated, err := filepath.EvalSymlinks(resolvedRoot); err == nil {
		resolvedRoot = evaluated
	}

	manifestPath := filepath.Join(resolvedRoot, RosterManifestFilename)
	raw, err := os.ReadFile(manifestPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, rosterManifestErrorf("roster package at %s is missing %s",
				resolvedRoot, RosterManifestFilename)
		}
		return nil, rosterManifestErrorf("%s: %s", manifestPath, err)
	}

	var document map[string]any
	if err := json.Unmarshal(raw, &document); err != nil {
		return nil, rosterManifestErrorf("%s: invalid JSON: %s", manifestPath, err)
	}

	// Every required field, named together. Reporting them one at a time makes
	// a half-written manifest a sequence of edit-and-retry rounds.
	var missing []string
	for _, field := range []string{
		"schema_version", "id", "version",
		"catalog", "routing", "role_root", "shared_policy_root",
	} {
		if _, present := document[field]; !present {
			missing = append(missing, field)
		}
	}
	if len(missing) > 0 {
		return nil, rosterManifestErrorf("%s: missing required field(s): %s",
			manifestPath, strings.Join(missing, ", "))
	}

	version, ok := jsonInteger(document["schema_version"])
	if !ok || version != SupportedRosterSchemaVersion {
		return nil, rosterManifestErrorf(
			"%s: unsupported schema_version %v; this selector supports %d",
			manifestPath, document["schema_version"], SupportedRosterSchemaVersion)
	}

	manifest := &RosterManifest{Root: resolvedRoot, ManifestPath: manifestPath}
	for _, field := range []struct {
		name      string
		target    *string
		directory bool
	}{
		{"catalog", &manifest.Catalog, false},
		{"routing", &manifest.Routing, false},
		{"role_root", &manifest.RoleRoot, true},
		{"shared_policy_root", &manifest.SharedPolicyRoot, true},
	} {
		resolved, err := rosterResource(resolvedRoot, document[field.name], field.name, field.directory)
		if err != nil {
			return nil, err
		}
		*field.target = resolved
	}
	for _, field := range []struct {
		name   string
		target *string
	}{{"id", &manifest.ID}, {"version", &manifest.Version}} {
		text, ok := document[field.name].(string)
		if !ok || text == "" {
			return nil, rosterManifestErrorf("%s: field %s must be a non-empty string",
				manifestPath, pythonRepr(field.name))
		}
		*field.target = text
	}
	return manifest, nil
}

// rosterResource resolves one declared path, rejecting anything that escapes
// root.
func rosterResource(root string, value any, field string, directory bool) (string, error) {
	text, ok := value.(string)
	if !ok || text == "" {
		return "", rosterManifestErrorf(
			"roster manifest field %s must be a non-empty relative path", pythonRepr(field))
	}
	// filepath.Join drops root when text is absolute, exactly as pathlib's `/`
	// does -- so an absolute value surfaces here as an escape rather than
	// needing its own guard. Left as the single check on purpose: a later
	// refactor that "simplifies" this into a prefix test on the raw string
	// would stop catching `..` segments.
	candidate := filepath.Join(root, text)
	if filepath.IsAbs(text) {
		candidate = filepath.Clean(text)
	}
	candidate, err := filepath.Abs(candidate)
	if err != nil {
		return "", rosterManifestErrorf("roster manifest field %s: %s", pythonRepr(field), err)
	}
	relative, err := filepath.Rel(root, candidate)
	if err != nil || relative == ".." || strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
		return "", rosterManifestErrorf(
			"roster manifest field %s escapes its manifest directory: %s",
			pythonRepr(field), pythonRepr(text))
	}

	info, err := os.Stat(candidate)
	if err != nil {
		if directory {
			return "", rosterManifestErrorf(
				"roster manifest field %s names a directory that does not exist: %s",
				pythonRepr(field), pythonRepr(text))
		}
		return "", rosterManifestErrorf(
			"roster manifest field %s names a file that does not exist: %s",
			pythonRepr(field), pythonRepr(text))
	}
	if directory && !info.IsDir() {
		return "", rosterManifestErrorf(
			"roster manifest field %s names a directory that does not exist: %s",
			pythonRepr(field), pythonRepr(text))
	}
	if !directory && info.IsDir() {
		return "", rosterManifestErrorf(
			"roster manifest field %s names a file that does not exist: %s",
			pythonRepr(field), pythonRepr(text))
	}
	return candidate, nil
}
